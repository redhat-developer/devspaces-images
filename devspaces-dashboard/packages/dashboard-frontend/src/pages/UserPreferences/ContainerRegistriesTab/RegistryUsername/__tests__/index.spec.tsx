/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Form } from '@patternfly/react-core';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { RegistryUsernameFormGroup } from '..';

describe('Registry Username Input', () => {
  const mockOnChange = jest.fn();

  function getComponent(username: string): React.ReactElement {
    return (
      <Form>
        <RegistryUsernameFormGroup username={username} onChange={mockOnChange} />
      </Form>
    );
  }

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should correctly render the component', () => {
    const component = getComponent('Attt');
    render(component);

    const input = screen.queryByRole('textbox');
    expect(input).toBeTruthy();

    const json = renderer.create(component).toJSON();

    expect(json).toMatchSnapshot();
  });

  it('should show default placeholder', () => {
    render(getComponent(''));

    const placeholder = screen.queryByPlaceholderText('Enter a username');
    expect(placeholder).toBeTruthy();
  });

  it('should correctly re-render the component', () => {
    const component = getComponent('Atest');
    const { rerender } = render(component);

    const input = screen.getByRole('textbox');
    expect(input).toHaveValue('Atest');

    rerender(getComponent('Atestregn'));

    expect(input).toHaveValue('Atestregn');
  });

  it('should fire onChange event', () => {
    const component = getComponent('Ate');
    render(component);

    const input = screen.getByRole('textbox');

    userEvent.type(input, 'st');

    expect(mockOnChange).toHaveBeenCalledWith('Ates', 'success');
    expect(mockOnChange).toHaveBeenCalledWith('Atest', 'success');
  });

  describe('validation', () => {
    it('should handle maximum value length', () => {
      const component = getComponent('https://testreg.com/test1');
      render(component);

      const input = screen.getByRole('textbox');

      const message = 'The username is too long. The maximum length is 100 characters.';
      const allowedUsername = 'a'.repeat(100);
      let label: HTMLElement | null;

      userEvent.clear(input);
      userEvent.type(input, allowedUsername);

      label = screen.queryByText(message);
      expect(label).toBeFalsy();

      const disallowedUsername = 'a'.repeat(101);

      userEvent.clear(input);
      userEvent.type(input, disallowedUsername);

      label = screen.queryByText(message);
      expect(label).toBeTruthy();
      expect(input).toBeInvalid();
    });
  });
});
