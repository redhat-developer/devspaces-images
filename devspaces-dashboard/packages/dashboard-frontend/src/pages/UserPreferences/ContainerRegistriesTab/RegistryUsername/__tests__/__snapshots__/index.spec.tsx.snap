// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registry Username Input should correctly render the component 1`] = `
<form
  className="pf-c-form"
  noValidate={true}
>
  <div
    className="pf-c-form__group"
    style={
      Object {
        "gridTemplateColumns": "80px",
        "minHeight": "65px",
      }
    }
  >
    <div
      className="pf-c-form__group-label"
    >
      <label
        className="pf-c-form__label"
        htmlFor="id-username-helper"
      >
        <span
          className="pf-c-form__label-text"
        >
          Username
        </span>
      </label>
       
    </div>
    <div
      className="pf-c-form__group-control"
    >
      <input
        aria-invalid={false}
        aria-label="Username input"
        className="pf-c-form-control"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Enter a username"
        readOnly={false}
        required={false}
        type="text"
        value="Attt"
      />
      
    </div>
  </div>
</form>
`;
